File: leetcode/uniquePaths2/UniquePaths2.py
# https://leetcode.com/problems/unique-paths-ii/submissions/
class UniquePaths2:
    def uniquePathsWithObstacles(self, obstacleGrid) -> int:
        N = len(obstacleGrid)
        M = len(obstacleGrid[0])
        matrix = [[0]*M for _ in range(N)]

        for row in range(N):
            if obstacleGrid[row][0] == 0:
                matrix[row][0] = 1
            else:
                matrix[row][0] = 0
                break

        for col in range(M):
            if obstacleGrid[0][col] == 0:
                matrix[0][col] = 1
            else:
                matrix[0][col] = 0
                break

        for row in range(1, N):
            for col in range(1, M):
                if obstacleGrid[row][col] == 0:
                    matrix[row][col] = matrix[row][col-1] + matrix[row-1][col]

        return matrix[N-1][M-1]

File: leetcode/uniquePaths2/UniquePaths2.java
package uniquePaths2;

// https://leetcode.com/problems/unique-paths-ii/
public class UniquePaths2 {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int N = obstacleGrid.length;
        int M = obstacleGrid[0].length;
        if (obstacleGrid[N-1][M-1] == 1) return 0;

        int[][] paths = new int[N][M];

        for (int i = 0; i < N; i++) {
            if (obstacleGrid[i][0] == 1) break;
            paths[i][0] = 1;
        }

        for (int i = 0; i < M; i++) {
            if (obstacleGrid[0][i] == 1) break;
            paths[0][i] = 1;
        }

        for (int row = 1; row < N; row++) {
            for (int col = 1; col < M; col++) {
                if (obstacleGrid[row][col] == 0) {
                    paths[row][col] = paths[row-1][col] + paths[row][col-1];
                }
            }
        }
        return paths[N-1][M-1];
    }
}
